### PR TITLE
feat: Better support for dataclasses

### LINKED ISSUE #233: bug: A dataclass field with init=False should not be a parameter
Here are failing tests.

```python
from griffe.tests import temporary_visited_module

code = """
from dataclasses import dataclass, field

@dataclass
class PointA:
    x: float
    y: float
    z: float = field(init=False)

@dataclass(init=False)
class PointB:
    x: float
    y: float

@dataclass(init=False)
class PointC:
    x: float
    y: float = field(init=True)  # init=True has no effect
"""

with temporary_visited_module(code) as module:
    paramsA = module["PointA"].parameters
    paramsB = module["PointB"].parameters
    paramsC = module["PointC"].parameters

    assert "z" not in paramsA
    assert "x" not in paramsB and "y" not in paramsB
    assert "x" not in paramsC and "y" not in paramsC
```

Then there are the perhaps more complicated cases to deal with statically where the `field` class is wrapped inside another function e.g.

```python
def no_init(default: T) -> T:
    """
    Set default value of a dataclass field that will not be __init__ed
    """
    return field(init=False, default=default)

@dataclass
class Point:
    x: float
    y: float = no_init(0)  # y is not a parameter
```

### PR DESCRIPTION
**Pull request by pawamoy**
Closes #233, #234, #238, #239.

TODO: 

- [x] remove `d.py`
- [x] add more tests (coming from mentioned issues)
- [ ] detect when dataclasses extension functionality was used, to issue a deprecation warning? or leave it enabled by default for eternity?

### CODE REVIEW

**Comment by pawamoy** [comment]
@has2k1 would you like to review? When squashing I'll add you as co-author since your issues were super helpful and I reused your code for tests :slightly_smiling_face:

File: src/griffe/extensions/base.py [hunk 1]
```diff
@@ -252,6 +252,9 @@ def __init__(self, *extensions: ExtensionType) -> None:
         self._extensions: list[Extension] = []
         self.add(*extensions)
 
+        # TODO: Deprecate and remove at some point?
+        self.add(*_load_extension("dataclasses"))
+
```
Comment (pawamoy) [inline src/griffe/extensions/base.py:257]: Oops, must remove this.

**Comment by has2k1** [comment]
Nice, looks good so far. I will have time to test it over the weekend.

File: src/griffe/extensions/base.py [hunk 2]
```diff
@@ -465,12 +468,14 @@ def load_extensions(exts: Sequence[str | dict[str, Any] | ExtensionType | type[E
...
     extensions = Extensions()
-    for extension in exts:
+    for extension in exts or ():
         ext = _load_extension(extension)
         if isinstance(ext, list):
             extensions.add(*ext)
         else:
             extensions.add(ext)
+    # TODO: Deprecate and remove at some point?
+    extensions.add(*_load_extension("dataclasses"))  # type: ignore[misc]
```
Comment (pawamoy) [inline src/griffe/extensions/base.py:478]: Should we add it first or last? Should we check if it is already added by the provided `exts`?

**Pull request merged**